dali/kernels/slice/slice_gpu.cuh

```diff
@@ -36,78 +36,123 @@ template <int Dims>
 struct SliceSampleDesc {
   void *__restrict__ out;
   const void *__restrict__ in;
+
+  const void *__restrict__ fill_values;
+  int channel_dim;
+
   TensorShape<Dims> in_strides;
   TensorShape<Dims> out_strides;
+
+  TensorShape<Dims> anchor;
+  TensorShape<Dims> in_shape;
 };
 
 struct BlockDesc {
   int sampleIdx;
-  size_t offset;
-  size_t size;
+  int64_t offset;
+  int64_t size;
 };
 
 template <int Dims, typename OutputType, typename InputType>
 __device__ void SliceFunc(OutputType *__restrict__ out, const InputType *__restrict__ in,
                           const int64_t *out_strides, const int64_t *in_strides,
+                          const int64_t *anchor, const int64_t *in_shape,
+                          const OutputType *__restrict__ fill_values, int channel_dim,
                           size_t offset, size_t block_end) {
-  if (Dims > 1 && out_strides[Dims-1] == in_strides[Dims-1]) {
-    const int NextDims = Dims > 1 ? Dims-1 : 1;
-    SliceFunc<NextDims>(out, in, out_strides, in_strides, offset, block_end);
+  if (Dims > 1 && out_strides[Dims - 1] == in_strides[Dims - 1] && anchor[Dims - 1] == 0 &&
+      channel_dim != Dims - 1) {
+    const int NextDims = Dims > 1 ? Dims - 1 : 1;
+    SliceFunc<NextDims>(out, in, out_strides, in_strides, anchor, in_shape, fill_values,
+                        channel_dim, offset, block_end);
     return;
   }
 
   for (; offset < block_end; offset += blockDim.x) {
-    size_t idx = offset;
-    size_t out_idx = idx;
-    size_t in_idx = 0;
+    int64_t idx = offset;
+    int64_t out_idx = idx;
+    int64_t in_idx = 0;
+    int i_c = 0;
+    bool out_of_range = false;
     for (int d = 0; d < Dims; d++) {
       int i_d = idx / out_strides[d];
-      idx %= static_cast<size_t>(out_strides[d]);
-      in_idx += i_d * static_cast<size_t>(in_strides[d]);
+      idx %= out_strides[d];
+      if (d == channel_dim)
+        i_c = i_d;
+
+      int in_i_d = anchor[d] + i_d;
+      out_of_range |= in_i_d < 0 || in_i_d >= in_shape[d];
+      if (!out_of_range)
+        in_idx += in_i_d * in_strides[d];
     }
     in_idx += idx;  // remaining dims have equal strides
-    out[out_idx] = clamp<OutputType>(in[in_idx]);
+    out[out_idx] = out_of_range ? fill_values[i_c] : clamp<OutputType>(in[in_idx]);
   }
 }
 
 template <typename OutputType, typename InputType, int Dims>
 __global__ void SliceKernel(const SliceSampleDesc<Dims> *samples, const BlockDesc *blocks) {
   int sampleIdx = blocks[blockIdx.x].sampleIdx;
-  size_t offset = blocks[blockIdx.x].offset + threadIdx.x;
-  size_t block_end = blocks[blockIdx.x].offset + blocks[blockIdx.x].size;
+  int64_t offset = blocks[blockIdx.x].offset + threadIdx.x;
```

Comment:
Why? Can it be negative?

Reply:
No, it was just a decision we had to use int and int64_t to simplify things (avoid having to cast back and forth all the time, template arguments not being picked automatically because the type differs, etc...)